File: packages/backend-node/src/domains/cells-windows.ts
import type {
  CellsWindowsApi,
  SpiceCharCell,
  SpiceDoubleCell,
  SpiceIntCell,
  SpiceWindow,
} from "@rybosome/tspice-backend-contract";
import {
  assertSpiceInt32,
  assertSpiceInt32NonNegative,
} from "@rybosome/tspice-backend-contract";
import { invariant } from "@rybosome/tspice-core";

import type { NativeAddon } from "../runtime/addon.js";

/** Create a {@link CellsWindowsApi} implementation backed by the native Node addon. */
export function createCellsWindowsApi(native: NativeAddon): CellsWindowsApi {
  return {
    ssize: (size, cell) => {
      assertSpiceInt32NonNegative(size, "ssize(size)");
      native.ssize(size, cell);
    },
    scard: (card, cell) => {
      assertSpiceInt32NonNegative(card, "scard(card)");
      native.scard(card, cell);
    },
    card: (cell) => {
      const out = native.card(cell);
      invariant(typeof out === "number", "Expected card() to return a number");
      return out;
    },
    size: (cell) => {
      const out = native.size(cell);
      invariant(typeof out === "number", "Expected size() to return a number");
      return out;
    },
    valid: (size, n, cell) => {
      assertSpiceInt32NonNegative(size, "valid(size)");
      assertSpiceInt32NonNegative(n, "valid(n)");
      native.valid(size, n, cell);
    },
    insrti: (item, cell) => {
      assertSpiceInt32(item, "insrti(item)");
      native.insrti(item, cell);
    },
    insrtd: (item, cell) => {
      native.insrtd(item, cell);
    },
    insrtc: (item, cell) => {
      native.insrtc(item, cell);
    },
    wninsd: (left, right, window) => {
      native.wninsd(left, right, window);
    },
    wncard: (window) => {
      const out = native.wncard(window);
      invariant(typeof out === "number", "Expected wncard() to return a number");
      return out;
    },
    wnfetd: (window, index) => {
      assertSpiceInt32NonNegative(index, "wnfetd(index)");
      const out = native.wnfetd(window, index);
      invariant(Array.isArray(out) && out.length === 2, "Expected wnfetd() to return [left,right]");
      invariant(typeof out[0] === "number" && typeof out[1] === "number", "Expected wnfetd() to return numbers");
      return [out[0], out[1]] as const;
    },
    wnvald: (size, n, window) => {
      assertSpiceInt32NonNegative(size, "wnvald(size)");
      assertSpiceInt32NonNegative(n, "wnvald(n)");
      native.wnvald(size, n, window);
    },
  };
}

File: packages/py-parity-checking/sidecars/spiceypy/src/py_parity_sidecar/handlers/cells_windows.py
from __future__ import annotations

from typing import Any

import spiceypy as sp
from spiceypy.utils.support_types import SPICECHAR_CELL, SPICEDOUBLE_CELL, SPICEINT_CELL

from ..models import (
    StepCellsWindowsCard,
    StepCellsWindowsInsrtc,
    StepCellsWindowsInsrtd,
    StepCellsWindowsInsrti,
    StepCellsWindowsScard,
    StepCellsWindowsSize,
    StepCellsWindowsSsize,
    StepCellsWindowsValid,
    StepCellsWindowsWncard,
    StepCellsWindowsWnfetd,
    StepCellsWindowsWninsd,
    StepCellsWindowsWnvald,
    StepOutput,
    WorkflowStep,
)
from ..runtime import SidecarRuntimeContext


def _get_or_create_int_cell(
    context: SidecarRuntimeContext,
    cell_id: str,
    *,
    max_cardinality: int,
) -> Any:
    int_cells = context.state.cellsWindows.intCells
    existing = int_cells.get(cell_id)
    if existing is not None:
        return existing

    if max_cardinality <= 0:
        raise ValueError("maxCardinality must be > 0")

    cell = SPICEINT_CELL(max_cardinality)
    int_cells[cell_id] = cell
    context.register_finalizer(
        f"cellsWindows.releaseIntCell:{cell_id}",
        lambda: context.state.cellsWindows.intCells.pop(cell_id, None),
    )
    return cell


def _require_int_cell(context: SidecarRuntimeContext, cell_id: str) -> Any:
    int_cells = context.state.cellsWindows.intCells
    cell = int_cells.get(cell_id)
    if cell is None:
        raise ValueError(f"Int cell does not exist: {cell_id}")
    return cell


def _get_or_create_double_cell(
    context: SidecarRuntimeContext,
    cell_id: str,
    *,
    max_cardinality: int,
) -> Any:
    double_cells = context.state.cellsWindows.doubleCells
    existing = double_cells.get(cell_id)
    if existing is not None:
        return existing

    if max_cardinality <= 0:
        raise ValueError("maxCardinality must be > 0")

    cell = SPICEDOUBLE_CELL(max_cardinality)
    double_cells[cell_id] = cell
    context.register_finalizer(
        f"cellsWindows.releaseDoubleCell:{cell_id}",
        lambda: context.state.cellsWindows.doubleCells.pop(cell_id, None),
    )
    return cell


def _require_double_cell(context: SidecarRuntimeContext, cell_id: str) -> Any:
    double_cells = context.state.cellsWindows.doubleCells
    cell = double_cells.get(cell_id)
    if cell is None:
        raise ValueError(f"Double cell does not exist: {cell_id}")
    return cell


def _get_or_create_char_cell(
    context: SidecarRuntimeContext,
    cell_id: str,
    *,
    max_cardinality: int,
    length: int,
) -> Any:
    char_cells = context.state.cellsWindows.charCells
    existing = char_cells.get(cell_id)
    if existing is not None:
        return existing

    if max_cardinality <= 0:
        raise ValueError("maxCardinality must be > 0")

    if length <= 0:
        raise ValueError("length must be > 0")

    cell = SPICECHAR_CELL(max_cardinality, length)
    char_cells[cell_id] = cell
    context.register_finalizer(
        f"cellsWindows.releaseCharCell:{cell_id}",
        lambda: context.state.cellsWindows.charCells.pop(cell_id, None),
    )
    return cell


def _require_char_cell(context: SidecarRuntimeContext, cell_id: str) -> Any:
    char_cells = context.state.cellsWindows.charCells
    cell = char_cells.get(cell_id)
    if cell is None:
        raise ValueError(f"Char cell does not exist: {cell_id}")
    return cell


def _get_or_create_window(context: SidecarRuntimeContext, window_id: str, *, max_intervals: int) -> Any:
    windows = context.state.cellsWindows.windows
    existing = windows.get(window_id)
    if existing is not None:
        return existing

    if max_intervals <= 0:
        raise ValueError("maxIntervals must be > 0")

    window = SPICEDOUBLE_CELL(max_intervals * 2)
    windows[window_id] = window
    context.register_finalizer(
        f"cellsWindows.releaseWindow:{window_id}",
        lambda: context.state.cellsWindows.windows.pop(window_id, None),
    )
    return window


def _require_window(context: SidecarRuntimeContext, window_id: str) -> Any:
    windows = context.state.cellsWindows.windows
    window = windows.get(window_id)
    if window is None:
        raise ValueError(f"Window does not exist: {window_id}")
    return window


def _require_target(context: SidecarRuntimeContext, target_kind: str, target_id: str) -> Any:
    if target_kind == "int":
        return _require_int_cell(context, target_id)

    if target_kind == "double":
        return _require_double_cell(context, target_id)

    if target_kind == "char":
        return _require_char_cell(context, target_id)

    if target_kind == "window":
        return _require_window(context, target_id)

    raise ValueError(f"Unsupported cells-windows targetKind: {target_kind}")


def run_cells_windows_step(step: WorkflowStep, context: SidecarRuntimeContext) -> StepOutput | None:
    if isinstance(step, StepCellsWindowsCard):
        target = _require_target(context, step.targetKind, step.targetId)
        return StepOutput(op=step.op, value=int(sp.card(target)))

    if isinstance(step, StepCellsWindowsInsrtc):
        cell = _get_or_create_char_cell(
            context,
            step.cellId,
            max_cardinality=step.maxCardinality or 8,
            length=step.length or 32,
        )
        sp.insrtc(step.item, cell)
        return StepOutput(op=step.op, value=None)

    if isinstance(step, StepCellsWindowsInsrtd):
        cell = _get_or_create_double_cell(
            context,
            step.cellId,
            max_cardinality=step.maxCardinality or 8,
        )
        sp.insrtd(step.item, cell)
        return StepOutput(op=step.op, value=None)

    if isinstance(step, StepCellsWindowsInsrti):
        cell = _get_or_create_int_cell(
            context,
            step.cellId,
            max_cardinality=step.maxCardinality or 8,
        )
        sp.insrti(step.item, cell)
        return StepOutput(op=step.op, value=None)

    if isinstance(step, StepCellsWindowsScard):
        target = _require_target(context, step.targetKind, step.targetId)
        sp.scard(step.card, target)
        return StepOutput(op=step.op, value=None)

    if isinstance(step, StepCellsWindowsSize):
        target = _require_target(context, step.targetKind, step.targetId)
        return StepOutput(op=step.op, value=int(sp.size(target)))

    if isinstance(step, StepCellsWindowsSsize):
        target = _require_target(context, step.targetKind, step.targetId)
        sp.ssize(step.size, target)
        return StepOutput(op=step.op, value=None)

    if isinstance(step, StepCellsWindowsValid):
        target = _require_target(context, step.targetKind, step.targetId)
        sp.valid(step.size, step.n, target)
        return StepOutput(op=step.op, value=None)

    if isinstance(step, StepCellsWindowsWncard):
        window = _require_window(context, step.windowId)
        return StepOutput(op=step.op, value=int(sp.wncard(window)))

    if isinstance(step, StepCellsWindowsWninsd):
        window = _get_or_create_window(context, step.windowId, max_intervals=step.maxIntervals or 8)
        sp.wninsd(step.left, step.right, window)
        return StepOutput(op=step.op, value=None)

    if isinstance(step, StepCellsWindowsWnfetd):
        window = _require_window(context, step.windowId)
        left, right = sp.wnfetd(window, step.index)
        return StepOutput(op=step.op, value={"left": float(left), "right": float(right)})

    if isinstance(step, StepCellsWindowsWnvald):
        window = _require_window(context, step.windowId)
        sp.wnvald(step.size, step.n, window)
        return StepOutput(op=step.op, value=None)

    return None
